[clusteragent/admission] Create a separate package for each webhook and define a common interface for them

davidor opened the pull request:
### What does this PR do?

Refactors the admission controller webhooks code.

This is a proposal to organize the code a bit. This should not change the behavior of any of the webhooks.

The diff is a bit big, so here are the changes to make the review easier:
- Each webhook is now in a dedicated package. That way, maintainers of each of them can focus mostly on what's inside the package. There's also a `common` package with helpers used in multiple webhooks.
- There's a new interface that the webhooks need to implement. The interface contains the information needed to set up the webhook (name, endpoint, whether it's enabled, etc.).
- The interface allows us to simplify the code in `cmd/cluster-agent/subcommands/start/command.go`, `pkg/clusteragent/admission/controllers/webhook/controller_v1.go`, and `pkg/clusteragent/admission/controllers/webhook/controller_v1beta1.go`, because we no longer need to set up things individually for each webhook.
- The rest is mostly making some functions public so that they can be used in the new packages.


### Describe how to test/QA your changes

Make sure that the webhooks still work. A quick test for them should be enough, just check that they inject something when they're enabled.

I believe that everything changed here is covered by unit tests.


- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

davidor committed:
[clusteragent/admission] Create a separate package for each webhook and a common interface for them

clamoriniere reviewed (commented):
here are some comments.
but overall, this PR looks good and will allow us to continue improving the organization of the webhook with some iterative PRs.

davidor committed:
[clusteragent/admission] Refactor webhook function

davidor commented on cmd/cluster-agent/admission/server.go:
```diff
@@ -38,7 +38,8 @@ import (
 
-type admissionFunc func(raw []byte, name string, namespace string, info *authenticationv1.UserInfo, dc dynamic.Interface, apiClient kubernetes.Interface) ([]byte, error)
+// Func is the function that runs the webhook logic
+type Func func(raw []byte, name string, namespace string, info *authenticationv1.UserInfo, dc dynamic.Interface, apiClient kubernetes.Interface) ([]byte, error)
```
I've pushed a new commit with this :+1:

davidor commented on pkg/clusteragent/admission/controllers/webhook/controller_base.go:
```diff
@@ -39,18 +45,62 @@ func NewController(client kubernetes.Interface, secretInformer coreinformers.Sec
+	MutateFunc() admission.Func
+}
+
+func mutatingWebhooks() []MutatingWebhook {
```
I prefer to do this in a separate PR if you don't mind because there are quite a few calls to the config and not all of them are in the init functions.

davidor commented:
@clamoriniere I've rebased the PR on top of main. The main change since you reviewed is that now the agent sidecar injection also follows the new structure defined. I've addressed one of your comments in a new commit.

davidor commented:
/merge
